### PR TITLE
Interaction system

### PR DESCRIPTION
**Pull request by Frogbeerr**
The interaction System is in place. For detailed information look into Assets/Player/InteractionContext.gd. There I documented how to use it.

Currently the selection system is broken since I have yet to integrate it into the interaction system properly. I'll work on that next.

### CODE REVIEW

File: Assets/Player/InteractionContext.gd [hunk 1]
```diff
@@ -0,0 +1,90 @@
+extends Node
+class_name InteractionContext
+"""This Node is a template for different behaiviors of the interaction system.
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
"""This Node is a template for different behaviors of the interaction system.
```

File: Assets/Player/InteractionContext.gd [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+It is meant to be extended and not to be instanced directly.
+PlayerCamera will keep track of one active InteractionContext at a time and will
+pass any unhandled InputEvents to it.
+E.g. a build menu could invoke a switch to it's own InteractionContext to handle
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
E.g. a build menu could invoke a switch to its own InteractionContext to handle
```

File: Assets/Player/InteractionContext.gd [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+the display of ghost buildings and later the placement of actual building.
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
the display of ghost buildings and later the placement of actual buildings.
```

File: Assets/Player/InteractionContext.gd [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+If a military unit is selected the system could swap to an InteractionContext
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
If a military unit is selected, the system could swap to an InteractionContext
```

File: Assets/Player/InteractionContext.gd [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+responsible for military commands, etc.
+These are hypothetical examples and provide no information about actual
+implementation.
+
+Each InteractionContext is responsible for setting inputs as handled themselves
+if so desired.
+
+Per default any InteractionContext can send two different signals:
+	switch_context - This signal causes the interaction system to switch its
+					 active InteractionContext to the one Provided as parameter.
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
					 active InteractionContext to the one provided as parameter.
```

File: Assets/Player/InteractionContext.gd [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	abort_context - This signals causes the interaction system to revert to the
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
	abort_context - This signal causes the interaction system to revert to the
```

File: Assets/Player/InteractionContext.gd [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+					default context (probably selection mode)
+
+The PoolStringArray 'valid_actions' contains the names of all possible
+action names. These are the same names as they would be used with
+InputEvent.is_action().
+Each of these actions will be routed to a method identified by it's name.
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
Each of these actions will be routed to a method identified by its name.
```

File: Assets/Player/InteractionContext.gd [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+The mehtod should be constructed as follows:
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
The method should be constructed as follows:
```

File: Assets/UI/Scripts/CityInfo.gd [hunk 9]
```diff
@@ -20,7 +20,7 @@ const FACTION_SETTLEMENT = [
 	preload("res://Assets/UI/Icons/Widgets/CityInfo/settlement_black.png")
 ]
 
-export(Global.Faction) var faction := 0 setget set_faction
+export(int) var faction := 0 setget set_faction
```
Comment (pingwindev) [inline Assets/UI/Scripts/CityInfo.gd]: This isn't necessary. `Global.Faction` is an enum.

File: Assets/World/Game.gd [hunk 10]
```diff
@@ -24,7 +24,7 @@ func _process(_delta: float) -> void:
 		start_game()
 
 # Notification test (press N within a game session)
-func _input(event: InputEvent):
+func unhandled_input(event: InputEvent)->void:
```
Comment (pingwindev) [inline Assets/World/Game.gd:27]:
```suggestion
func unhandled_input(event: InputEvent) -> void:
```

File: Assets/Player/PlayerCamera.gd [hunk 11]
```diff
@@ -12,18 +13,15 @@ var start_sel_pos = Vector2()
 
 var player = null
 
+# Variables for the interaction system
+export (NodePath) var default_interaction_context
+var active_context : InteractionContext
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
var active_context: InteractionContext
```

**Review by pingwindev — COMMENTED** [review]
I will check the code later in the editor, so just skimming it here for obvious and stylistic issues.

By the way, you can mark a PR as a "Draft PR" in the dropdown below. Please use this as long as it's WIP.

**Comment by LinuxDonald** [comment]
@artism90 please final review and merge please

File: Assets/Player/InteractionContext.gd [hunk 12]
```diff
@@ -0,0 +1,92 @@
+extends Node
+class_name InteractionContext
+"""This Node is a template for different behaiviors of the interaction system.
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]: We don't use docstrings. The default comment style in GDScript is via `#` with a space in between like `# This is a comment.`.

File: Assets/Player/InteractionContext.gd [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+	func _on_mouse_motion(
+		event: InputEventMouseMotion,
+		target: Node,
+		position: Vector3
+		) -> void
+All InputEventMouseMotion events will be routed to this function. This can
+be used to update hover effects, the cursor, etc.
+"""
+
+
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd:45]: Only one line space before/after functions.
```suggestion

```

File: Assets/Player/InteractionContext.gd [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+signal switch_context
+signal abort_context
+
+onready var _p := get_parent()
+
+export (String) var _context_name = "Basic Interaction Context"
+export (PoolStringArray) var valid_actions = ["main_command"]
+
+func interact(event:InputEvent, target:Node, position:Vector3) -> void:
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
func interact(event: InputEvent, target: Node, position: Vector3) -> void:
```

File: Assets/Player/InteractionContext.gd [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+be used to update hover effects, the cursor, etc.
+"""
+
+
+signal switch_context
+signal abort_context
+
+onready var _p := get_parent()
+
+export (String) var _context_name = "Basic Interaction Context"
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
export(String) var _context_name = "Basic Interaction Context"
```

File: Assets/Player/InteractionContext.gd [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+export (PoolStringArray) var valid_actions = ["main_command"]
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
export(PoolStringArray) var valid_actions = ["main_command"]
```

File: Assets/Player/InteractionContext.gd [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+func interact(event:InputEvent, target:Node, position:Vector3) -> void:
+	if event.is_action_type():
+		for action in valid_actions:
+			if event.is_action(action):
+				var function : String = make_function_name(
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
				var function: String = make_function_name(
```

File: Assets/Player/InteractionContext.gd [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+					action, event.is_action_pressed(action))
+				if self.has_method(function):
+					self.call(function, target, position)
+					return
+	elif event is InputEventMouseMotion:
+		_on_mouse_motion(event, target, position)
+		return
+
+func make_function_name(action:String, pressed:bool=true) -> String:
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
func make_function_name(action: String, pressed: bool = true) -> String:
```

File: Assets/Player/InteractionContext.gd [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+	var func_name : String = action.replace(" ", "_")
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
	var func_name: String = action.replace(" ", "_")
```

File: Assets/Player/InteractionContext.gd [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+func get_valid_actions() -> PoolStringArray:
+	return valid_actions
+
+func _on_enter() -> void:
+	print_debug("InteractionContext %s entered" % _context_name)
+
+func _on_exit() -> void:
+	print_debug("InteractionContext %s left" % _context_name)
+
+func _on_mouse_motion(event:InputEventMouseMotion, target:Node, position:Vector3)->void:
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]:
```suggestion
func _on_mouse_motion(event: InputEventMouseMotion, target: Node, position: Vector3) -> void:
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 21]
```diff
@@ -0,0 +1,72 @@
+extends InteractionContext
+
+signal set_selection
+
+export (int) var selection_tolerance = 10
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
export(int) var selection_tolerance = 10
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+var sel_pos_start : Vector3
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
var sel_pos_start: Vector3
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+var sel_pos_end : Vector3
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
var sel_pos_end: Vector3
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+
+
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion

```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+func _on_ia_alt_command_pressed(target : Node, position : Vector3) -> void:
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
func _on_ia_alt_command_pressed(target: Node, position: Vector3) -> void:
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+	"""Start selection"""
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	# Start selection
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+	print_debug("Start selection")
+	sel_pos_start = position
+	_p.set_sel_pos_start(position)
+	_p.show()
+
+func _on_ia_alt_command_released(target : Node, position : Vector3) -> void:
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
func _on_ia_alt_command_released(target: Node, position: Vector3) -> void:
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+	"""End selection"""
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	# End selection
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+	print_debug("End selection")
+	sel_pos_end = position
+	var selection : Array
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	var selection: Array
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+	var top_left : Vector2 = get_viewport().get_camera().unproject_position(sel_pos_start)
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	var top_left: Vector2 = get_viewport().get_camera().unproject_position(sel_pos_start)
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+	var bot_right : Vector2 = get_viewport().get_camera().unproject_position(sel_pos_end)
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	var bot_right: Vector2 = get_viewport().get_camera().unproject_position(sel_pos_end)
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+	if top_left.distance_to(bot_right) <= selection_tolerance:
+		if target.is_in_group("units"):
+			selection = [target]
+	else:
+		selection = get_selected_units(top_left, bot_right)
+	_p.hide()
+	emit_signal("set_selection", selection)
+
+func _on_ia_main_command_pressed(target : Node, position : Vector3) -> void:
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
func _on_ia_main_command_pressed(target: Node, position: Vector3) -> void:
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 33]
```diff
@@ -0,0 +1,72 @@
...
+	"""Abort selection"""
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	# Abort selection
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+	print_debug("Abort selection")
+	_p.hide()
+
+func get_selected_units(top_left : Vector2, bot_right : Vector2) -> Array:
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
func get_selected_units(top_left: Vector2, bot_right: Vector2) -> Array:
```

File: Assets/Player/InteractionContexts/SelectionContext.gd [hunk 35]
```diff
@@ -0,0 +1,72 @@
...
+	if top_left.x > bot_right.x:
+		var tmp = top_left.x
+		top_left.x = bot_right.x
+		bot_right.x = tmp
+	if top_left.y > bot_right.y:
+		var tmp = top_left.y
+		top_left.y = bot_right.y
+		bot_right.y = tmp
+	var box = Rect2(top_left, bot_right - top_left)
+	var selected_units : Array =  []
```
Comment (pingwindev) [inline Assets/Player/InteractionContexts/SelectionContext.gd]:
```suggestion
	var selected_units: Array = []
```

File: Assets/Player/PlayerCamera.gd [hunk 36]
```diff
@@ -12,18 +13,15 @@ var start_sel_pos = Vector2()
...
-#func _ready() -> void:
-#	get_tree().call_group(
-#		"billboard",
-#		"update_offset",
-#		_rotation_y.rotation_degrees.y)
-#	get_tree().call_group(
-#		"billboard",
-#		"recalculate_directions",
-#		_rotation_y.rotation_degrees.y)
+func _ready():
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
func _ready() -> void:
```

File: Assets/Player/PlayerCamera.gd [hunk 37]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
-			box_selected_units.append(unit)
-	return box_selected_units
 
 func raycast_from_mouse(m_pos: Vector2, collision_mask: int) -> Dictionary:
 	var ray_start = _camera.project_ray_origin(m_pos)
 	var ray_end = ray_start + _camera.project_ray_normal(m_pos) * RAY_LENGTH
 	var space_state = get_world().direct_space_state
 	return space_state.intersect_ray(ray_start, ray_end, [], collision_mask)
+
+
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion

```

File: Assets/Player/PlayerCamera.gd [hunk 38]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+### INTERACTION SYSTEM ###
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
# Interaction system
```

File: Assets/Player/PlayerCamera.gd [hunk 39]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+func switch_context(new_context : InteractionContext) -> void:
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
func switch_context(new_context: InteractionContext) -> void:
```

File: Assets/Player/PlayerCamera.gd [hunk 40]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+	if not active_context.is_connected("abort_context", self, "abort_context"):
+		active_context.connect("abort_context", self, "abort_context")
+	if not active_context.is_connected("switch_context", self, "switch_context"):
+		active_context.connect("switch_context", self, "switch_context")
+	active_context._on_enter()
+
+func abort_context() -> void:
+	switch_context(get_node(default_interaction_context))
+
+
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion

```

File: Assets/Player/PlayerCamera.gd [hunk 41]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+func _unhandled_input(event: InputEvent) -> void:
+	var m_pos := get_viewport().get_mouse_position()
+	var target := raycast_from_mouse(m_pos, 1)
+	var target_object : Node
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
	var target_object: Node
```

File: Assets/Player/PlayerCamera.gd [hunk 42]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+	var target_pos : Vector3
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
	var target_pos: Vector3
```

File: Assets/Player/PlayerCamera.gd [hunk 43]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+	if target:
+		target_object = (target["collider"] as Node).get_parent()
+		target_pos = (target["position"] as Vector3)
+	active_context.interact(event, target_object, target_pos)
+
+
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion

```

File: Assets/Player/PlayerCamera.gd [hunk 44]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+func set_selection(new_sel : Array) -> void:
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
func set_selection(new_sel: Array) -> void:
```

File: Assets/Player/PlayerHud.gd [hunk 45]
```diff
@@ -26,3 +26,7 @@ func raise_notification(message_type: int, message_text: String) -> void:
 		messages.add_child(message)
 #	else:
 #		queued_messages.append()
+
+func _gui_input(event:InputEvent)->void:
```
Comment (pingwindev) [inline Assets/Player/PlayerHud.gd]:
```suggestion
func _gui_input(event: InputEvent) -> void:
```

File: Assets/UI/Scripts/SelectionBox.gd [hunk 46]
```diff
@@ -1,18 +1,26 @@
...
-var is_visible = false
-var m_pos = Vector2()
-var start_sel_pos = Vector2() # first click position
 
-const sel_box_col = Color(1, 1, 1)
-const sel_box_line_width = 1
+onready var _p = get_parent()
+var sel_pos_start = Vector3() # first click position
+
+export (Color) var line_color = Color(1, 1, 1)
```
Comment (pingwindev) [inline Assets/UI/Scripts/SelectionBox.gd]:
```suggestion
export(Color) var line_color = Color(1, 1, 1)
```

File: Assets/UI/Scripts/SelectionBox.gd [hunk 47]
```diff
@@ -1,18 +1,26 @@
...
+export (int) var line_width = 1
```
Comment (pingwindev) [inline Assets/UI/Scripts/SelectionBox.gd]:
```suggestion
export(int) var line_width = 1
```

File: Assets/UI/Scripts/SelectionBox.gd [hunk 48]
```diff
@@ -1,18 +1,26 @@
...
 
 func _draw() -> void:
-	if is_visible and start_sel_pos != m_pos:
-		draw_line(start_sel_pos, Vector2(m_pos.x, start_sel_pos.y), sel_box_col, sel_box_line_width) # upper right
-		draw_line(start_sel_pos, Vector2(start_sel_pos.x, m_pos.y), sel_box_col, sel_box_line_width) # lower left
-		draw_line(m_pos, Vector2(m_pos.x, start_sel_pos.y), sel_box_col, sel_box_line_width) # upper left
-		draw_line(m_pos, Vector2(start_sel_pos.x, m_pos.y), sel_box_col, sel_box_line_width) # lower right
+	if is_visible():
+		var m_pos = get_viewport().get_mouse_position()
+		var top_left : Vector2 = get_viewport().get_camera().unproject_position(sel_pos_start)
```
Comment (pingwindev) [inline Assets/UI/Scripts/SelectionBox.gd]:
```suggestion
		var top_left: Vector2 = get_viewport().get_camera().unproject_position(sel_pos_start)
```

File: Assets/UI/Scripts/SelectionBox.gd [hunk 49]
```diff
@@ -1,18 +1,26 @@
...
+		if top_left != m_pos:
+			draw_line(top_left, Vector2(m_pos.x, top_left.y), line_color, line_width) # upper right
+			draw_line(top_left, Vector2(top_left.x, m_pos.y), line_color, line_width) # lower left
+			draw_line(m_pos, Vector2(m_pos.x, top_left.y), line_color, line_width) # upper left
+			draw_line(m_pos, Vector2(top_left.x, m_pos.y), line_color, line_width) # lower right
 
 func _process(_delta: float) -> void:
 	update()
+
+func set_sel_pos_start(tl : Vector3) -> void:
```
Comment (pingwindev) [inline Assets/UI/Scripts/SelectionBox.gd]:
```suggestion
func set_sel_pos_start(tl: Vector3) -> void:
```

File: Assets/UI/Scripts/ShipMenuTabWidget.gd [hunk 50]
```diff
@@ -4,5 +4,5 @@ class_name ShipMenuTabWidget
 
 onready var faction_indicator = $WidgetDetail/Body/ShipMenu/MarginContainer/FactionIndicator
 
-func _ready() -> void:
-	faction_indicator.texture = Global.FACTION_FLAGS[Global.faction]
+#func _ready() -> void:
+#	faction_indicator.texture = Global.FACTION_FLAGS[Global.faction]
```
Comment (pingwindev) [inline Assets/UI/Scripts/ShipMenuTabWidget.gd]: Why is this out commented? When I revert it back within your PR, it works fine.
```suggestion
func _ready() -> void:
	faction_indicator.texture = Global.FACTION_FLAGS[Global.faction]
```

**Review by pingwindev — CHANGES_REQUESTED** [review]
I went again through the code for style issues, since I've spot of lot of them while studying it, so above are all occurrences that need to be addressed so far and a few other remarks.

The code style generally follows the [official GDScript style guide](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_styleguide.html).

Exceptions:
`extends`, then `class_name`.
Generally **one** line break where it says **two** in the official guide.

File: Assets/Player/PlayerCamera.gd [hunk 51]
```diff
@@ -41,84 +39,58 @@ func _process(_delta: float) -> void:
...
+	"""update selection"""
```
Comment (pingwindev) [inline Assets/Player/PlayerCamera.gd]:
```suggestion
	# Update selection
```

File: Assets/UI/Scripts/SelectionBox.gd [hunk 52]
```diff
@@ -1,18 +1,26 @@
 extends Control
 
-var is_visible = false
-var m_pos = Vector2()
-var start_sel_pos = Vector2() # first click position
 
-const sel_box_col = Color(1, 1, 1)
-const sel_box_line_width = 1
+onready var _p = get_parent()
```
Comment (pingwindev) [inline Assets/UI/Scripts/SelectionBox.gd]: Is this a leftover? It doesn't do anything right now.
```suggestion

```

File: Assets/Player/InteractionContext.gd [hunk 53]
```diff
@@ -0,0 +1,92 @@
...
+		) -> void
+All InputEventMouseMotion events will be routed to this function. This can
+be used to update hover effects, the cursor, etc.
+"""
+
+
+signal switch_context
+signal abort_context
+
+onready var _p := get_parent()
```
Comment (pingwindev) [inline Assets/Player/InteractionContext.gd]: As a rule of thumb, better to follow an explicit, descriptive naming:
```suggestion
onready var _parent := get_parent()
```

**Review by aaronfranke — CHANGES_REQUESTED** [review]
For this to be approved, you will need to rebase, and then address the above feedback. If you need help rebasing, please see [this documentation article](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html), or watch [this video](https://www.youtube.com/watch?v=0XKy8YpP5Ww).

**Comment by pingwindev** [comment]
_Note: This was merged too early. I'll commit a few things directly later to fix remaining bugs. Please no merging without prior approvement._

Applied all suggestions/remaining changes since OP is inactive. Also squashed several pointless commits in between into the main one.
Side note: Existing player ship movement functionality is broken per this PR and has to be re-implemented within the interaction system as a `ShipMovementContext` or the like.

**Pull request merged**